libavformat/rtcenc.c

```diff
@@ -0,0 +1,1548 @@
+ *       a=fingerprint:sha-256 68:DD:7A:95:27:BD:0A:99:F4:7A:83:21:2F:50:15:2A:1D:1F:8A:D8:96:24:42:2D:A1:83:99:BF:F1:E2:11:A2
+ *       a=setup:passive
+ *       a=mid:0
+ *       a=recvonly
+ *       a=rtcp-mux
+ *       a=rtcp-rsize
+ *       a=rtpmap:111 opus/48000/2
+ *       a=candidate:0 1 udp 2130706431 172.20.10.7 8000 typ host generation 0
+ *
+ *       m=video 9 UDP/TLS/RTP/SAVPF 106
+ *       c=IN IP4 0.0.0.0
+ *       a=ice-ufrag:ex9061f9
+ *       a=ice-pwd:bi8k19m9n836187b00d1gm3946234w85
+ *       a=fingerprint:sha-256 68:DD:7A:95:27:BD:0A:99:F4:7A:83:21:2F:50:15:2A:1D:1F:8A:D8:96:24:42:2D:A1:83:99:BF:F1:E2:11:A2
+ *       a=setup:passive
+ *       a=mid:1
+ *       a=recvonly
+ *       a=rtcp-mux
+ *       a=rtcp-rsize
+ *       a=rtpmap:106 H264/90000
+ *       a=fmtp:106 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01e
+ *       a=candidate:0 1 udp 2130706431 172.20.10.7 8000 typ host generation 0
+ *
+ * @return 0 if OK, AVERROR_xxx on error
+ */
+static int exchange_sdp(AVFormatContext *s)
+{
+    int ret;
+    char buf[MAX_URL_SIZE];
+    RTCContext *rtc = s->priv_data;
+    /* The URL context is an HTTP transport layer for the WHIP protocol. */
+    URLContext *whip_uc = NULL;
+
+    char *tmp = av_mallocz(MAX_SDP_SIZE);
+    if (!tmp) {
+        av_log(s, AV_LOG_ERROR, "Failed to alloc answer: %s\n", s->url);
+        return AVERROR(ENOMEM);
+    }
+
+    ret = ffurl_alloc(&whip_uc, s->url, AVIO_FLAG_READ_WRITE, &s->interrupt_callback);
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to alloc HTTP context: %s\n", s->url);
+        goto end;
+    }
+
+    snprintf(buf, sizeof(buf),
+             "Cache-Control: no-cache\r\n"
+             "Content-Type: application/sdp\r\n");
+    av_opt_set(whip_uc->priv_data, "headers", buf, 0);
+    av_opt_set(whip_uc->priv_data, "chunked_post", "0", 0);
+    av_opt_set_bin(whip_uc->priv_data, "post_data", rtc->sdp_offer, (int)strlen(rtc->sdp_offer), 0);
+
+    ret = ffurl_connect(whip_uc, NULL);
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to request url=%s, offer: %s\n", s->url, rtc->sdp_offer);
+        goto end;
+    }
+
+    while (1) {
+        ret = ffurl_read(whip_uc, buf, sizeof(buf));
+        if (ret == AVERROR_EOF) {
+            /* Reset the error because we read all response as answer util EOF. */
+            ret = 0;
+            break;
+        }
+        if (ret <= 0) {
+            av_log(s, AV_LOG_ERROR, "Failed to read response from url=%s, offer is %s, answer is %s\n",
+                s->url, rtc->sdp_offer, rtc->sdp_answer);
+            goto end;
+        }
+
+        ret = av_strlcatf(tmp, MAX_SDP_SIZE, "%.*s", ret, buf);
+        if (ret >= MAX_SDP_SIZE) {
+            av_log(s, AV_LOG_ERROR, "Answer %d exceed max size %d, %s\n", ret, MAX_SDP_SIZE, tmp);
+            ret = AVERROR(EIO);
+            goto end;
+        }
+    }
+
+    rtc->sdp_answer = av_strdup(tmp);
```

Comment:
add memory alloc failed check and give failed message when it failed.
other code block for memory alloc should have same operation.

Reply:
Roger that. All similar issues are fixed by https://github.com/ossrs/ffmpeg-webrtc/pull/1/commits/e5bbc9164f12d521ccd240a8d930a56465ce2f59